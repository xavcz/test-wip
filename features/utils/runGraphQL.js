import { graphql } from 'graphql';
import decomposeOperation from './decomposeOperation';

export default async ({
  schema,
  operation,
  context,
  variables,
  showLogs = false,
}) => {
  try {
    // get the operation string & AST
    const { source, information } = decomposeOperation(operation);

    const result = await graphql({
      schema,
      source,
      contextValue: context,
      variableValues: variables,
    });

    const { data, errors = [] } = result;

    if (showLogs) {
      console.log('\n// operation\n', operation);
      console.log('\n// operation result\n', data);
      console.log('\n// operation errors\n', errors);
    }

    return { value: data[information.selectionName], error: errors[0] };
  } catch (err) {
    throw new Error(err);
  }
};
