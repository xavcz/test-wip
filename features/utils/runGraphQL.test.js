import runGraphQL from './runGraphQL';
import { makeExecutableSchema, addMockFunctionsToSchema } from 'graphql-tools';

describe('runGraphQL', () => {
  const schema = makeExecutableSchema({
    typeDefs: `
      type Query { say: String }
      type Mutation { addOneTo(number: Int!): Int }
    `,
  });
  addMockFunctionsToSchema({ schema });

  const context = {};

  it('should run an operation against a schema', async () => {
    const result = await runGraphQL({
      schema,
      context,
      operation: 'query hey { say }',
    });

    expect(result).toMatchSnapshot();
  });

  it('should reject the promise on invalid operation', () => {
    expect(
      runGraphQL({
        schema,
        context,
        operation: '{ say }',
      })
    ).rejects.toBeDefined();

    expect(
      runGraphQL({
        schema,
        context,
        operation: 'mutation { addOneTo(number: 20) }',
      })
    ).rejects.toBeDefined();
  });
});
