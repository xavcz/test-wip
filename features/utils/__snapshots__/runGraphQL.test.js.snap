// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runGraphQL should run an operation against a schema 1`] = `
Object {
  "error": undefined,
  "value": "Hello World",
}
`;
